danceradvent/public/articles/2020/23-asynchronous-promises.pod

```diff
@@ -0,0 +1,526 @@
+L<previous article|https://advent.perldancer.org/2020/21>, you might see
+how our code begins to shift to the right with growing callbacks. Each
+async code requires providing anonymous subroutines that will get
+triggered once the async request (whether API request or DB call)
+returns. Our code will start to look like an arrow.
+
+    use experimental qw< postderef signatures >;
+
+    do_this_async( sub ($foo) {
+        do_this_other_thing_async( $foo, sub ($bar) {
+            foreach my $baz ( $bar->@* ) {
+                do_more_async( $baz, sub ($quux) {
+                    ...
+                });
+            }
+        });
+    });
+
+Eventually, this can become quite unwieldy. Promises is a pattern
+that can help us tame our code. L<Promises> is an exceptional implementation
+of this pattern, so we will use it.
+
+=head2 So what are Promises?
+
+Promises are a simple mechanism for turning that arrow pattern above
+into a clear straight chain where each callback is stored in an object
+using a method, and the eventual value can be used in another callback,
+again stored in the same object using a method.
+
+L<Promises> also provides us with comfortable method chaining..
+
+Let's take the above code and rewrite it using the Promises syntax:
+
+    use experimental qw< postderef signatures >;
+
+    do_this_async()->then( sub ($foo) {
+        return do_this_other_thing_async($foo)
+    })->then( sub ($bar) {
+        my @promises = do_more_async($_)
+            for $bar->@*;
+
+        return collect(@promises);
+    })->then( sub (@results) {
+        ...
+    });
+
+Assuming that C<do_this_async> returns a L<Promise> object,
+we can call C<then> to tack on a callback for the result that
+this method will eventually retrieve.
+
+We can then call another method (C<call_this_other_thing_async>),
+which will also return a Promise, to which we tack another callback
+that will use the value we retrieved from the previous method.
+
+Notice everything here so far had been done with callbacks. We don't
+execute anything except the initial C<do_this_async>. The rest simply
+registers callbacks. When it all gets executed at the end, this smart
+chaining will cause them to be triggered in the right order, connecting
+them together.
+
+Next, we call a method C<do_more_async> which returns a Promise again,
+for each of our input. We collect those in a variable and call the
+C<Promises> function C<collect> which creates a single promise from a
+set of them. We can then tack a callback for it using C<then>.
+
+Of course, we will need to have these functions return Promise objects,
+so you'll get to see that.
+
+=head2 Okay, so how do we rewrite this?
+
+You might not have fully understood how Promises - that's okay. But even if
+you did, you might not know how to move our previous code to this. No matter
+which category you fall in (maybe both?), we're now going to do it together.
+
+=head3 Update our imports
+
+package CovidStats::Promises;
```

Comment:
Need to indent this line so it formats as code.

Reply:
Fixed.